### PR TITLE
test: add stop/start pipeline test

### PR DESCRIPTION
**Pull request by keqiaozhang**
This script intends to test stop/start aplay/arecord on each pipeline.

Signed-off-by: Zhang Keqiao <keqiao.zhang@linux.intel.com>

### CODE REVIEW

**Comment by xiulipan** [comment]
@keqiaozhang can you change the test case name with signal stop start? I think stop start is very confusing.

**Comment by keqiaozhang** [comment]
Changed the case name to "check-signal-stop-start.sh".

File: test-case/check-signal-stop-start.sh [hunk 1]
```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
+
+##
+## Case Name: check-stop-start
```
Comment (xiulipan) [inline test-case/check-signal-stop-start.sh]: case name also need update

File: test-case/check-signal-stop-start.sh [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+FILE['playback,both']='/dev/zero'
+CMD['capture,both']='arecord'
+FILE['capture,both']='/dev/null'
+
+func_stop_start_pipeline()
+{
+    count=1
+    while(true)
+    do
+        ps -ef |grep "$pid" |grep -v grep
```
Comment (xiulipan) [inline test-case/check-signal-stop-start.sh]: can we use some other check logic? like `kill -0 $pid`?
@Bin-QA any suggestion?

File: test-case/check-signal-stop-start.sh [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            dlogi "Stop/start count: $count"
+            # stop the pipeline
+            kill -SIGSTOP $pid
+            sleep $interval
+            # start the pipeline
+            kill -SIGCONT $pid
+            sleep $interval
+            let count++
+        else
+        break # aplay/arecord is finished, terminate stop/start test
```
Comment (xiulipan) [inline test-case/check-signal-stop-start.sh]: intend here.

File: test-case/check-signal-stop-start.sh [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+declare -A CMD FILE
+CMD['playback,both']='aplay'
+FILE['playback,both']='/dev/zero'
+CMD['capture,both']='arecord'
+FILE['capture,both']='/dev/null'
+
+func_stop_start_pipeline()
+{
+    count=1
```
Comment (xiulipan) [inline test-case/check-signal-stop-start.sh]: can we make this `local`

File: test-case/check-signal-stop-start.sh [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        pcm=$(func_pipeline_parse_value $idx pcm)
+        pipeline_type=$(func_pipeline_parse_value $idx "type")
+
+        dlogi "Testing: run stop/start test on PCM:$pcm,$pipeline_type. Interval time: $interval"
+        dlogc "${CMD[$1]}" -D$dev -r $rate -c $channel -f $fmt -d $duration "${FILE[$1]}" -q &
+        "${CMD[$1]}" -D$dev -r $rate -c $channel -f $fmt -d $duration "${FILE[$1]}" -q & pid=$!
+        # do stop/start test
+        func_stop_start_pipeline
+        # check aplay/arecord return value
+        wait $pid
```
Comment (xiulipan) [inline test-case/check-signal-stop-start.sh]: @keqiaozhang @Bin-QA some question here.
We will only exit the `func_stop_start_pipeline` when aplay/arecord exit, can we wait for it after the exit?

Comment (Bin-QA) [reply test-case/check-signal-stop-start.sh]: I think the function loop can direct load form here look like this, here `wait` use to query process exit code,.
```
while [ "$(ps -p $pid)"]
do
            dlogi "Stop/start count: $count"
            # stop the pipeline
            kill -SIGSTOP $pid
            sleep $interval
            # start the pipeline
            kill -SIGCONT $pid
            sleep $interval
            let count++
done
wait $pid
```

File: test-case/check-signal-stop-start.sh [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            dloge "$pipeline_type on $pcm failed."
+            exit 1
+        fi
+    done
+}
+
+func_test_pipeline_with_type "playback,both"
+func_test_pipeline_with_type "capture,both"
+
+sof-kernel-log-check.sh $KERNEL_LAST_LINE
```
Comment (Bin-QA) [inline test-case/check-signal-stop-start.sh]: @keqiaozhang, you need to load this function `func_lib_setup_kernel_last_line` to get `$KERNEL_LAST_LINE` value

File: test-case/check-signal-stop-start.sh (hunk 1 above)
Comment (keqiaozhang) [reply test-case/check-signal-stop-start.sh]: solved.

File: test-case/check-signal-stop-start.sh (hunk 2 above)
Comment (keqiaozhang) [reply test-case/check-signal-stop-start.sh]: @Bin-QA  good suggestion. Updated.

File: test-case/check-signal-stop-start.sh (hunk 6 above)
Comment (keqiaozhang) [reply test-case/check-signal-stop-start.sh]: Updated.

**Comment by xiulipan** [comment]
@keqiaozhang one concern here, it seems we always allow user to select playback/capture for the test. Would it keep in that style or merge the playback and capture inside the test case?

File: test-case/check-signal-stop-start.sh [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+OPT_OPT_lst['d']='duration' OPT_DESC_lst['d']='duration time'
+OPT_PARM_lst['d']=1         OPT_VALUE_lst['d']=10
+
+OPT_OPT_lst['s']='sof-logger'   OPT_DESC_lst['s']="Open sof-logger trace the data will store at $LOG_ROOT"
+OPT_PARM_lst['s']=0             OPT_VALUE_lst['s']=1
+
+func_opt_parse_option $*
+
+tplg=${OPT_VALUE_lst['t']}
+count=${OPT_VALUE_lst['c']}
```
Comment (wenqingfu) [inline test-case/check-signal-stop-start.sh]: 'c' is not defined. There is no use of it, either.  Please enable the support of $count number of start/stop.

In the meantime, do you need to specify duration in this test? can you let it run indefinitely and kill aplay/arecord once you reached the $count?

**Pull request merged**